dvl-a50/dvl.py

```diff
@@ -175,12 +171,72 @@ def set_should_send(self, should_send):
         self.should_send = should_send
         self.save_settings()
 
-    def set_gps_origin(self, lat, lon):
+    @staticmethod
+    def longitude_scale(lat: float):
+        """
+        from https://github.com/ArduPilot/ardupilot/blob/Sub-4.1/libraries/AP_Common/Location.cpp#L325
+        """
+        scale = math.cos(math.radians(lat))
+        return max(scale, 0.01)
+
+    def lat_lng_to_NE_XY_cm (self, lat: float, lon: float) -> List[float]:
+        """
+        From https://github.com/ArduPilot/ardupilot/blob/Sub-4.1/libraries/AP_Common/Location.cpp#L206
+        """
+        x = (lat-self.origin[0]) * LATLON_TO_CM
+        y =  self.longitude_scale((lat + self.origin[0])/2) * LATLON_TO_CM * (lon-self.origin[1])
+        return [x, y]
+
+    def has_origin_set(self) -> bool:
+        try:
+            old_time = self.mav.get_float("/GPS_GLOBAL_ORIGIN/message/time_usec")
+            if old_time == float('nan'):
+                logger.warning("Unable to read current time for GPS_GLOBAL_ORIGIN, using 0")
+        except Exception as e:
+            logger.warning(f"Unable to read current time for GPS_GLOBAL_ORIGIN, using 0: {e}")
```

Comment:
as `using 0` is not supposed to set something here ?

Reply:
damn, I just introduced this bug yesterday, nice catch